File: server/src/main/kotlin/xyz/upperlevel/hgame/server/Player.kt
package xyz.upperlevel.hgame.server

import io.netty.channel.Channel
import xyz.upperlevel.hgame.matchmaking.InvitePacket
import xyz.upperlevel.hgame.matchmaking.InvitePacketType
import java.lang.IllegalStateException
import java.lang.RuntimeException

class Player(val channel: Channel) {
    var name: String? = null

    val isLoginDone: Boolean
        get() = name != null

    var lobby: Lobby? = null
    var character: String? = null
    var ready = false

    var relayChannel: Channel? = null


    // The invite system is quite strange
    // when you invite someone the invite is valid until you
    // join another invite, when you do the invites you sent get invalidated.
    // To simulate this behaviour we create two sets with the invites that
    // we sent and the ones we received, when a player joins a lobby
    // that he didn't create his sentInvites gets cleared
    // and the receivedInvites of the players that were in there
    // are modified accordingly.
    val receivedInvites = HashSet<Player>()
    val sentInvites = HashSet<Player>()


    fun onLogin(name: String) {
        if (isLoginDone) throw IllegalStateException("Login already succeed")
        this.name = name
    }

    fun sendInvite(to: Player) {
        sentInvites.add(to)
        to.onInviteReceived(this)
    }

    fun onInviteReceived(inviter: Player) {
        receivedInvites.add(inviter)
        channel.writeAndFlush(InvitePacket(InvitePacketType.INVITE_RECEIVED, inviter.name!!))
    }

    fun acceptInvite(from: Player, lobbyRegistry: LobbyRegistry): String? {
        if (from !in receivedInvites) {
            return "Invite expired"
        }

        val lobby = from.lobby ?: lobbyRegistry.create(from)

        return lobby.onJoin(this)
    }

    fun invalidateSentInvites() {
        sentInvites.forEach {
            it.receivedInvites.remove(this)
        }
        sentInvites.clear()
    }
}

File: server/src/player.ts
import {MatchmakingPacket} from "../../common/src/matchmaking/protocol"
import * as ws from 'ws';
import {Lobby, LobbyRegistry} from "./lobby";


export class Player {
    playerRegistry: PlayerRegistry;

    private _name?: string;
    lobby?: Lobby;
    character?: string;
    ready: boolean = false;

    receivedInvites: Set<Player> = new Set();
    sentInvites: Set<Player> = new Set();

    readonly socket: ws;
    relaySocket?: ws;

    constructor(playerRegistry: PlayerRegistry, socket: ws) {
        this.playerRegistry = playerRegistry;
        this.socket = socket;
    }

    get name(): string {
        if (this._name == null) {
            throw new Error("login not done yet")
        }
        return this._name
    }

    get isLoginDone(): boolean {
        return this._name != null
    }

    onLogin(name: string): boolean {
        if (!this.playerRegistry.onLogin(this, name)) return false;

        if (this.isLoginDone) throw new Error("Login already succeded");

        this._name = name;
        return true;
    }

    sendInvite(to: Player) {
        this.sentInvites.add(to);
        to.onInviteReceived(this);
    }

    private onInviteReceived(from: Player) {
        this.receivedInvites.add(from);

        this.sendPacket({
            type: "invite",
            kind: "INVITE_RECEIVED",
            player: from.name
        })
    }

    sendPacket(packet: MatchmakingPacket) {
        this.socket.send(JSON.stringify(packet))
    }

    acceptInvite(from: Player, lobbyRegistry: LobbyRegistry): string | undefined {
        if (!this.receivedInvites.has(from)) {
            return "Invite expired"
        }
        let lobby = from.lobby;
        if (lobby == null) {
            lobby = lobbyRegistry.create(from)
        }
        return lobby.onJoin(this)
    }

    invalidateSentInvites() {
        this.sentInvites.forEach((invited) => {
            invited.receivedInvites.delete(this);
        });
        this.sentInvites.clear();
    }

    onDisconnect() {
        this.invalidateSentInvites();

        this.playerRegistry.onDisconnect(this);

        if (this.lobby != null) {
            this.lobby.onQuit(this)
        }
    }
}

export class PlayerRegistry {
    players: { [name: string]: Player } = {};

    onConnect(socket: ws): Player {
        return new Player(this, socket);
    }

    onLogin(player: Player, name: string): boolean {
        if (name in this.players) {
            return false;
        }
        this.players[name] = player;
        return true;
    }

    onDisconnect(player: Player) {
        if (player.name != null) return;

        delete this.players[player.name];
    }

    getByName(name: string): Player | undefined {
        return this.players[name]
    }
}


